File: release/serve.mjs
// Same job as serve.py, for a machine that has Node but not Python.
// Node's standard library only, no install, no package.json.
//
// Usage: node serve.mjs [port]

import { createServer } from 'node:http';
import { createReadStream, statSync } from 'node:fs';
import { extname, resolve, sep } from 'node:path';
import { fileURLToPath } from 'node:url';
import os from 'node:os';

const ROOT = fileURLToPath(new URL('.', import.meta.url));
const ROOT_RESOLVED = resolve(ROOT);

// .mjs must be a JavaScript type or the browser refuses to run the pdf.js
// worker as a module, and no page ever renders.
const TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
};

const server = createServer((req, res) => {
  const path = decodeURIComponent(new URL(req.url, 'http://localhost').pathname);
  // resolve() collapses any ".." and the prefix check then rejects anything
  // that climbed out of this folder, so a crafted URL cannot read the disk.
  const file = resolve(ROOT, path === '/' ? 'index.html' : path.slice(1));
  if (file !== ROOT_RESOLVED && !file.startsWith(ROOT_RESOLVED + sep)) {
    res.writeHead(403, { 'Content-Type': 'text/plain' });
    return res.end('Forbidden');
  }
  try {
    if (!statSync(file).isFile()) throw new Error('not a file');
  } catch {
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    return res.end('Not found');
  }
  res.writeHead(200, {
    'Content-Type': TYPES[extname(file).toLowerCase()] ?? 'application/octet-stream',
    'Cache-Control': 'no-store',
  });
  createReadStream(file).pipe(res);
});

const lan = process.argv.includes('--lan');
const port = Number(process.argv.filter((a) => a !== '--lan')[2] ?? 8000);

// Loopback unless --lan is passed. With --lan anyone who can reach this
// machine on this port gets the app, and nothing here checks who is asking.
const host = lan ? '0.0.0.0' : '127.0.0.1';
server.listen(port, host, () => {
  if (lan) {
    console.log('*** --lan: this is now reachable by other machines on the network. ***');
    console.log('*** There is no authentication. Check this is allowed before use. ***');
    console.log(`Others use: http://${os.hostname()}:${port}/`);
  }
  console.log(`PDF Requirement Designer running at http://127.0.0.1:${port}/`);
  console.log('Leave this window open. Press Ctrl+C to stop.');
});
server.on('error', (err) => {
  console.error(err.code === 'EADDRINUSE' ? `Port ${port} is busy, try: node serve.mjs 8080` : err.message);
  process.exit(1);
});

File: release/serve.py
"""Serves this folder over HTTP so the designer can run.

The app cannot be opened straight off the filesystem: it loads as an ES module
and reads two JSON files, and browsers block both over file://. So it needs an
HTTP origin, which is all this does. Python's standard library only, no install.

The .mjs mapping below is not optional. Python maps .mjs to text/plain, and a
browser refuses to run a module script served as text/plain, which would leave
pdf.js with no worker and no page ever rendering.

Usage: python serve.py [port] [--lan]

Without --lan this listens on 127.0.0.1 and is reachable from this machine
only. Read the warning under --lan before using it.
"""

import http.server
import mimetypes
import socket
import socketserver
import sys
import threading
import webbrowser

mimetypes.add_type("text/javascript", ".mjs")
mimetypes.add_type("text/javascript", ".js")
mimetypes.add_type("application/json", ".json")


class Handler(http.server.SimpleHTTPRequestHandler):
    def end_headers(self):
        # The field lists are meant to be swapped without a rebuild, so make
        # sure a stale copy is never served back.
        self.send_header("Cache-Control", "no-store")
        super().end_headers()

    def log_message(self, fmt, *args):
        pass  # Keep the window readable; errors still surface in the browser.


def free_port(host, preferred):
    for port in [preferred] + list(range(preferred + 1, preferred + 20)):
        with socket.socket() as s:
            try:
                s.bind((host, port))
                return port
            except OSError:
                continue
    raise SystemExit("No free port in range.")


def main():
    args = [a for a in sys.argv[1:] if a != "--lan"]
    lan = "--lan" in sys.argv
    preferred = int(args[0]) if args else 8000

    host = "0.0.0.0" if lan else "127.0.0.1"
    port = free_port(host, preferred)

    if lan:
        # Anyone who can reach this machine on this port gets the app, with no
        # login of any kind. Nothing here checks who is asking.
        print("*** --lan: this is now reachable by other machines on the network. ***")
        print("*** There is no authentication. Check this is allowed before use. ***")
        print(f"Others use: http://{socket.gethostname()}:{port}/")

    url = f"http://127.0.0.1:{port}/"
    with socketserver.TCPServer((host, port), Handler) as httpd:
        print(f"PDF Requirement Designer running at {url}")
        print("Leave this window open. Press Ctrl+C to stop.")
        threading.Timer(0.5, lambda: webbrowser.open(url)).start()
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\nStopped.")


if __name__ == "__main__":
    main()
